src/main/java/com/google/sps/servlets/GenerateServlet.java

```diff
@@ -28,14 +32,143 @@
-    Gson gson = new Gson();
     HashSet<String> clickedFilters =
-        gson.fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+        new Gson().fromJson(request.getParameter(FILTER_ARRAY), HashSet.class);
+
+    // Create fake destinations TODO: get destinations from datastore
+    ArrayList<Destination> allDestinations = createFakeDestinations();
+
+    // Create arrays for each category of filter
+    HashSet<String> userPlaces = new HashSet(clickedFilters);
+    HashSet<String> userDifficultyStrings = new HashSet(clickedFilters);
+
+    // Retain only the filters that user chose (for each category)
+    userPlaces.retainAll(Arrays.asList(allPlaces));
+    userDifficultyStrings.retainAll(Arrays.asList(allDifficulties));
+
+    // Convert difficulty level strings to Destination.Obscurity
+    HashSet<Destination.Obscurity> userDifficultyLevels =
+        convertStringsToEnum(userDifficultyStrings);
+
+    // Filter
+    Set<Destination> filteredDestinations =
+        filter(allDestinations, userPlaces, userDifficultyLevels);
+
+    writeToDataStore(filteredDestinations);
 
     response.setContentType("text/html;");
     response.getWriter().println(clickedFilters);
   }
+
+  /* Takes strings of difficulty level and converts into Destination.Obscurity objects. */
+  public HashSet<Destination.Obscurity> convertStringsToEnum(
+      HashSet<String> userDifficultyStrings) {
+    HashSet<Destination.Obscurity> userDifficultyLevels = new HashSet();
+    if (userDifficultyStrings.isEmpty()) {
+      userDifficultyLevels.add(Destination.Obscurity.EASY);
+      userDifficultyLevels.add(Destination.Obscurity.MEDIUM);
+      userDifficultyLevels.add(Destination.Obscurity.HARD);
+    } else {
+      for (String userLevel : userDifficultyStrings) {
+        switch (userLevel) {
+          case "Easy":
+            userDifficultyLevels.add(Destination.Obscurity.EASY);
+            break;
+          case "Medium":
+            userDifficultyLevels.add(Destination.Obscurity.MEDIUM);
+            break;
+          case "Hard":
+            userDifficultyLevels.add(Destination.Obscurity.HARD);
+            break;
+        }
+      }
+    }
+    return userDifficultyLevels;
+  }
+
+  public Set<Destination> filter(
+      ArrayList<Destination> allDestinations,
+      HashSet<String> userPlaces,
+      HashSet<Destination.Obscurity> userDifficultyLevels) {
+    // Filter by place
+    Set<Destination> filteredDestinations =
+        allDestinations.stream()
+            .filter(destination -> userPlaces.contains(destination.getCity()))
+            .map(destination -> destination)
+            .collect(Collectors.toSet());
+    // Filter by difficulty
+    filteredDestinations =
+        filteredDestinations.stream()
+            .filter(destination -> userDifficultyLevels.contains(destination.getDifficulty()))
+            .map(destination -> destination)
+            .collect(Collectors.toSet());
```

Comment:
Two things:

1) This can be combined!  You want to filter by city _and_ difficulty here, so you can just specify that in the filter function.

2) The map method will transform the left hand side into the right hand side, i.e. if you did map(destination -> destination.getName()) it would turn your set of Destinations into a set of Strings.  Since you're not transforming destination, you don't need the map() part.

```suggestion
    Set<Destination> filteredDestinations =
        allDestinations.stream()
            .filter(destination -> userPlaces.contains(destination.getCity()) && userDifficultyLevels.contains(destination.getDifficulty()))
            .collect(Collectors.toSet());
```